trustymail/trustymail.py

```diff
@@ -63,6 +68,76 @@ def mx_scan(domain):
         handle_error("[MX]", domain, error)
 
 
+def starttls_scan(domain, timeout):
+    """
+    Scan a domain to see if it sends mail and supports STARTTLS.
+
+    Scan a domain to see if it sends mail.  If the domain does send
+    mail, a further check will be done to see if it supports STARTTLS.
+    All results are stored inside the Domain object that is passed in
+    as a parameter.
+
+    Parameters
+    ----------
+    domain : Domain
+        The Domain to be tested.
+        
+    timeout : int
+        The SMTP connection timeout in seconds.  (Default is 5.)
+    """
+    for mail_server in domain.mail_servers:
+        for port in SMTP_PORTS:
+            smtp_connection = smtplib.SMTP(timeout=timeout)
+            server_and_port = mail_server + ":" + str(port)
+            logging.debug("Testing " + server_and_port + " for STARTTLS support")
+            # Try to connect.  This will tell us if something is
+            # listening.
+            try:
+                smtp_connection.connect(mail_server, port)
+            except (socket.timeout, smtplib.SMTPConnectError, smtplib.SMTPServerDisconnected, ConnectionRefusedError, OSError) as error:
+                handle_error("[STARTTLS]", domain, error)
+                tmp = server_and_port + "-" + str(False)
+                domain.sends_mail.append(tmp)
+                domain.starttls.append(tmp)
+                continue
+
+            # Now try to say hello.  This will tell us if the thing
+            # that is listening is an SMTP server.
+            try:
+                smtp_connection.ehlo_or_helo_if_needed()
+            except (smtplib.SMTPHeloError, smtplib.SMTPServerDisconnected) as error:
+                handle_error("[STARTTLS]", domain, error)
+                tmp = server_and_port + "-" + str(False)
+                domain.sends_mail.append(tmp)
+                domain.starttls.append(tmp)
+                # smtplib freaks out if you call quit on a non-open
+                # connection
+                try:
+                    smtp_connection.quit()
+                except smtplib.SMTPServerDisconnected as error2:
+                    handle_error("[STARTTLS]", domain, error2)
+                    
+                continue
+
+            # We were able to connect to the server and say hello, so
+            # it must send mail.
+            domain.sends_mail.append(server_and_port + "-" + str(True))
```

Comment:
We may want to play with the verbiage here; "sends mail" may be fudgy. I don't necessarily think that technical correctness should go ahead of user friendliness, but just that "sends mail" may be too general. 

Port 25 is used for mail transmission while port 587 is for _mail submission_ (as is port 465, unofficially). The standards track RFC (4409) that defines [mail submission](https://tools.ietf.org/html/rfc4409#section-3.1), and especially the [BCP on mail submission](https://tools.ietf.org/html/rfc5068#section-2), describe how mail transit and submission are different. Indeed, the [STARTTLS RFC](https://tools.ietf.org/html/rfc3207#section-4.3) calls out the submission port as being different from (and benefitting especially because) STARTTLS is operating on the submission port.

Reply:
How about "supports SMTP"?

Reply:
👍 yep, that works.